libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -63,119 +78,167 @@ void framing::abort(const error& reason) {
   // Parse header.
-  detail::rfc6455::header hdr;
-  auto hdr_bytes = detail::rfc6455::decode_header(buffer, hdr);
+  auto hdr_bytes = detail::rfc6455::decode_header(buffer, hdr_);
   if (hdr_bytes < 0) {
     CAF_LOG_DEBUG("decoded malformed data: hdr_bytes < 0");
     abort_and_shutdown(sec::protocol_error,
                        "negative header size on WebSocket connection");
     return -1;
   }
-  if (hdr_bytes == 0) {
-    // Wait for more input.
+  // Wait for more input.
+  if (hdr_bytes == 0)
     return 0;
-  }
-  if (detail::rfc6455::is_control_frame(hdr.opcode)) {
+  // Check the length of the payload
+  if (detail::rfc6455::is_control_frame(hdr_.opcode)) {
     // control frames can only have payload up to 125 bytes
-    if (hdr.payload_len > 125) {
+    if (hdr_.payload_len > 125) {
       CAF_LOG_DEBUG("WebSocket control frame payload exceeds allowed size");
       abort_and_shutdown(sec::protocol_error, "WebSocket control frame payload "
                                               "exceeds allowed size");
       return -1;
     }
   } else {
     // Make sure the entire frame (including header) fits into max_frame_size.
-    if (hdr.payload_len >= max_frame_size - static_cast<size_t>(hdr_bytes)) {
+    if (hdr_.payload_len >= max_frame_size - static_cast<size_t>(hdr_bytes)) {
       CAF_LOG_DEBUG("WebSocket frame too large");
       abort_and_shutdown(sec::protocol_error, "WebSocket frame too large");
       return -1;
     }
     // Reject any payload that exceeds max_frame_size. This covers assembled
-    //  payloads as well by including payload_buf_.
-    if (payload_buf_.size() + hdr.payload_len > max_frame_size) {
+    // payloads as well by including payload_buf_.
+    if (payload_buf_.size() + hdr_.payload_len > max_frame_size) {
       CAF_LOG_DEBUG("fragmented WebSocket payload exceeds maximum size");
       abort_and_shutdown(sec::protocol_error, "fragmented WebSocket payload "
                                               "exceeds maximum size");
       return -1;
     }
   }
-  // Wait for more data if necessary.
-  size_t frame_size = hdr_bytes + hdr.payload_len;
-  if (buffer.size() < frame_size) {
-    down_->configure_read(receive_policy::exactly(frame_size));
-    return 0;
-  }
-  // Decode frame.
-  auto payload_len = static_cast<size_t>(hdr.payload_len);
-  auto payload = buffer.subspan(hdr_bytes, payload_len);
-  if (hdr.mask_key != 0) {
-    detail::rfc6455::mask_data(hdr.mask_key, payload);
+  return hdr_bytes;
+}
+
+ptrdiff_t framing::consume_payload(byte_span buffer, byte_span delta) {
+  // Calculate at what point of the received buffer the delta payload begins.
+  auto offset = static_cast<ptrdiff_t>(buffer.size() - delta.size());
+  // We read frame_size at most. This can leave unprocessed input.
+  auto payload = buffer.first(std::min(buffer.size(), hdr_.payload_len));
+  // Unmask the arrived data.
+  if (hdr_.mask_key != 0)
+    detail::rfc6455::mask_data(hdr_.mask_key, payload, offset);
```

Comment:
Why can we receive more data then the payload len here? I would expect to see simply `detail::rfc6455::mask_data(hdr_.mask_key, buffer, offset);` and have `consume_header` prepare the transport with `up_to(hdr.payload_len)`. Is there a bug preventing this?

Reply:
Basically at the start of processing, and at every new call to consume, we set the receive policy to `ut_to(2048);`. But I agree, I can work around this now and no need for `consume_payload` to be concerned about this.